File: day06/day06-aws-vpc-cs/day06-aws-vpc-cs/SubnetUtils.cs
using System;
using System.Collections.Generic;
using System.Net;

namespace day06_aws_vpc_cs
{
    public class SubnetUtils
    {
        private static IPAddress IntToIpAddress(int ip)
        {
            var ipBytes = new byte[4];
            for (var i = 0; i < 4; i++)
            {
                ipBytes[i] = (byte)(ip >> (24 - (8 * i)) & 0xFF);
            }

            return new IPAddress(ipBytes);
        }

        public static List<string> SplitSubnets(string baseCidr, int newPrefixLength)
        {
            var subnets = new List<string>();

            var baseCidrIp = IPAddress.Parse(baseCidr.Split('/')[0]);
            var baseCidrPrefixLength = int.Parse(baseCidr.Split('/')[1]);

            // calculate total ips to int type
            var totalIps = (int)Math.Pow(2, 32 - baseCidrPrefixLength) - 1;


            var addressBytes = baseCidrIp.GetAddressBytes();
            int ip = 0;
            for (int i = 0; i < addressBytes.Length; i++)
            {
                ip += addressBytes[i] << (24 - (8 * i));
            }

            var lastIpAddressInt = ip + totalIps;

            var incrementIp = (int)Math.Pow(2, 32 - newPrefixLength);

            while (ip < lastIpAddressInt)
            {
                subnets.Add(IntToIpAddress(ip) + "/" + newPrefixLength);
                ip += incrementIp;
            }

            return subnets;
        }
    }
}

File: day06/day06-aws-vpc-java/src/main/java/tw/raccoon/pulumi/day06/SubnetUtils.java
package tw.raccoon.pulumi.day06;

import java.net.InetAddress;
import java.net.UnknownHostException;
import java.util.ArrayList;
import java.util.List;

public class SubnetUtils {

    private static InetAddress intToIpAddress(int ip) {
        byte[] ipBytes = new byte[4];
        for (int i = 0; i < 4; i++) {
            ipBytes[i] = (byte) ((ip >> (24 - (8 * i))) & 0xFF);
        }

        try {
            return InetAddress.getByAddress(ipBytes);
        } catch (UnknownHostException e) {
            e.printStackTrace();
            return null;
        }

    }

    public static List<String> splitSubnets(String baseCidr, int newPrefixLength) {
        List<String> subnets = new ArrayList<>();

        String[] parts = baseCidr.split("/");
        InetAddress baseCidrIp = null;
        try {
            baseCidrIp = InetAddress.getByName(parts[0]);
        } catch (UnknownHostException e) {
            throw new RuntimeException(e);
        }
        int baseCidrPrefixLength = Integer.parseInt(parts[1]);

        // Calculate total IPs to int type
        int totalIps = (int) Math.pow(2, 32 - baseCidrPrefixLength) - 1;

        byte[] addressBytes = baseCidrIp.getAddress();
        int ip = 0;
        for (int i = 0; i < addressBytes.length; i++) {
            ip += (addressBytes[i] & 0xFF) << (24 - (8 * i));
        }

        int lastIpAddressInt = ip + totalIps;
        int incrementIp = (int) Math.pow(2, 32 - newPrefixLength);

        while (ip < lastIpAddressInt) {
            subnets.add(intToIpAddress(ip).getHostAddress() + "/" + newPrefixLength);
            ip += incrementIp;
        }

        return subnets;
    }

}
